winit-wayland/src/seat/pointer/mod.rs

```diff
@@ -401,23 +401,16 @@ impl Default for WinitPointerDataInner {
 }
 
 /// Convert the Wayland button into winit.
-fn wayland_button_to_winit(button: u32) -> MouseButton {
+fn wayland_button_to_winit(button: u32) -> ButtonSource {
     // These values are coming from <linux/input-event-codes.h>.
-    const BTN_LEFT: u32 = 0x110;
-    const BTN_RIGHT: u32 = 0x111;
-    const BTN_MIDDLE: u32 = 0x112;
-    const BTN_SIDE: u32 = 0x113;
-    const BTN_EXTRA: u32 = 0x114;
-    const BTN_FORWARD: u32 = 0x115;
-    const BTN_BACK: u32 = 0x116;
-
-    match button {
-        BTN_LEFT => MouseButton::Left,
-        BTN_RIGHT => MouseButton::Right,
-        BTN_MIDDLE => MouseButton::Middle,
-        BTN_BACK | BTN_SIDE => MouseButton::Back,
-        BTN_FORWARD | BTN_EXTRA => MouseButton::Forward,
-        button => MouseButton::Other(button as u16),
+    const BTN_MOUSE: u32 = 0x110;
+    const BTN_JOYSTICK: u32 = 0x120;
+
+    if (BTN_MOUSE..BTN_JOYSTICK).contains(&button) {
+        // Mapping orders match
+        MouseButton((button - BTN_MOUSE) as u8).into()
+    } else {
+        ButtonSource::Unknown(button as u16)
     }
```

Comment:
This goes beyond the Wayland specification. I haven't seen any evidence that codes outside this range would be produced from mouse input, but maybe if a mouse has more than 16 buttons? Maybe there is some way buttons on a mouse could be mapped to other codes?

If anyone does have evidence that Wayland can produce codes outside this range (`0x110..=0x11F`) I'd like to see it.

Reply:
I mean, it's possible, I guess, but not right now. Like we can use `u16` for button just in case, since it doesn't really matter, I guess, just so we don't have a breaking change in the future if the problem appears?

Reply:
I put this in `ButtonSource::Unknown` instead of within `MouseButton` for a reason. Allowing it to contain undocumented `u16` codes defeats that reason.

I would prefer to revisit this when we have evidence of such values, then map those to values within `MouseButton`. I view changes to `Unknown` codes as non-breaking (well, maybe not appropriate for patch releases), while changes to `MouseButton` codes would be breaking changes.